File: quickSort/quick_sort.go
package quickSort

type List struct {
	Values []int
}

func (list *List) Sort() *List {
	sorted := quickSort(list.Values, 0, len(list.Values)-1)
	return &List{Values: sorted}
}

func quickSort(list []int, low int, high int) []int {
	if low > high {
		return list
	}

	p := partition(list, low, high)
	quickSort(list, low, p-1)
	quickSort(list, p+1, high)

	return list
}

func partition(list []int, low, high int) int {
	pivot := list[high]

	for i := low; i < high; i++ {
		if list[i] < pivot {
			list[i], list[low] = list[low], list[i]
			low++
		}
	}

	list[low], list[high] = list[high], list[low]

	return low
}

File: quickSort/quickSort.ts
function sort(arr: number[], lo: number, hi: number) {
    if (hi <= lo) return;
    const pivot = partition(arr, lo, hi);
    sort(arr, lo, pivot-1);
    sort(arr, pivot+1, hi);
    return arr;
}

function partition(arr: number[], lo: number, hi: number): number {
    let left = lo;
    let right = hi + 1;

    while (true) {
        while (arr[++left] < arr[lo])  {
            if(left === hi) break;
        }
        while (arr[lo] < arr[--right]) {
            if(right === lo) break;
        }

        if (left >= right) break;

        swap(arr, left, right);
    }


    // swap j with partitioning item
    swap(arr, lo, right)

    // return index of item now know to be in place
    return right;
}

function swap(arr: number[], i: number, j: number) {
    const tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp;
}

const arr = [19, 10, 30, 2, 4, 20, 1, 5];
console.log(`before sort:`);
console.log(arr);
const sorted = sort(arr, 0, arr.length-1)
console.log(`after sort:`);
console.log(sorted);
